### PR TITLE
use owner references instead of timestamps to determine build/build pod ownership

### PR DESCRIPTION
**Pull request by jim-minter**
Follow up from https://bugzilla.redhat.com/show_bug.cgi?id=1547551

### CODE REVIEW

**Comment by jim-minter** [comment]
@csrwng if you have a moment to comment on this PR it'd be appreciated.  @bparees and @gabemontero too.  I think we might have discussed this in the past, but I don't recall what the outcome was.  Perhaps back then it was "don't touch it unless something comes up in the future?"

In https://bugzilla.redhat.com/show_bug.cgi?id=1547551 a problem was encountered where the build controller rejected its own build pod because of timestamp differences, ultimately caused by faulty time synchronisation between the masters in the cluster in question (ntp not switched on).

I'm wondering if we can tighten up the code in this area.

So far this PR is a naive change to use the pre-existing OwnerReference (including UID) instead of timestamps as the factor that links build pods and builds.

However, I'm a bit in doubt about a couple of wider aspects of the code - added three questions as code comments.  Feedback would be appreciated.

I'm wondering whether this falls under "we added extra code for an unlikely corner case, but it doesn't actually quite work correctly and there are no tests" ?

File: pkg/build/controller/build/build_controller.go [hunk 1]
```diff
@@ -881,12 +883,24 @@ func (bc *BuildController) createBuildPod(build *buildapi.Build) (*buildUpdate,
...
+			// Error state.
 			existingPod, err := bc.podClient.Pods(build.Namespace).Get(buildPod.Name, metav1.GetOptions{})
-			if err == nil && existingPod.CreationTimestamp.Before(&build.CreationTimestamp) {
+			if err == nil && !strategy.HasOwnerReference(existingPod, build) {
 				update = transitionToPhase(buildapi.BuildPhaseError, buildapi.StatusReasonBuildPodExists, buildapi.StatusMessageBuildPodExists)
 				return update, nil
 			}
+			// QUESTION 1: isn't it a mistake to be returning nil, nil here in
+			// the case that err != nil?  Shouldn't we return nil, err in that
+			// case?
```
Comment (csrwng) [inline pkg/build/controller/build/build_controller.go]: I would say that yes it's a mistake to return nil, nil in case err != nil ... we want to retry handling this build.

File: pkg/build/controller/build/build_controller.go [hunk 2]
```diff
@@ -881,12 +883,24 @@ func (bc *BuildController) createBuildPod(build *buildapi.Build) (*buildUpdate,
...
+
+			// QUESTION 2: if we got here, and found a Pod that we know we
+			// previously created, presumably that implies we failed to Update()
+			// the build previously.  In that case, returning nil, nil here
+			// again fails to Update() the Build correctly.  Don't we then lose
+			// information like the push secret, etc. that we try to stash
+			// below?  Isn't that wrong? Should we / can we effectively jump to
+			// line 913 in this case (refactoring instead of goto)?
```
Comment (csrwng) [inline pkg/build/controller/build/build_controller.go]: I would say yes

File: pkg/build/controller/build/build_controller.go [hunk 3]
```diff
@@ -425,12 +425,14 @@ func (bc *BuildController) cancelBuild(build *buildapi.Build) (*buildUpdate, err
 func (bc *BuildController) handleNewBuild(build *buildapi.Build, pod *v1.Pod) (*buildUpdate, error) {
 	// If a pod was found, and it was created after the build was created, it
 	// means that the build is active and its status should be updated
+	// QUESTION 3: if we fix question 2 below, can we remove this entire stanza,
+	// and wouldn't that be the right thing to do?
```
Comment (csrwng) [inline pkg/build/controller/build/build_controller.go]: So if you already know that a pod exists, why not handle it here instead of checking policy, and then trying to create it?

File: pkg/build/controller/strategy/util.go [hunk 4]
```diff
@@ -269,19 +270,35 @@ func getPodLabels(build *buildapi.Build) map[string]string {
...
 	}
 }
 
+func setOwnerReference(pod *v1.Pod, build *buildapi.Build) {
+	pod.OwnerReferences = []metav1.OwnerReference{makeOwnerReference(build)}
+}
+
+// HasOwnerReference returns true if the build pod has an OwnerReference to the
+// build.
+func HasOwnerReference(pod *v1.Pod, build *buildapi.Build) bool {
```
Comment (gabemontero) [inline pkg/build/controller/strategy/util.go]: How about renaming the method to `HaveACommonOwnerReference`

File: pkg/build/controller/build/build_controller.go (hunk 3 above)
Comment (jim-minter) [reply pkg/build/controller/build/build_controller.go]: Because you know that you didn't successfully complete the work in the current state, therefore you shouldn't advance to the next.  The implementation related to handling and advancing the state should be in one place for legibility and maintainability.  And precisely one thing only should determine whether the work in the given state is complete.

**Comment by jim-minter** [comment]
Thanks for looking this over @csrwng!

File: pkg/build/controller/build/build_controller.go (hunk 2 above)
Comment (gabemontero) [reply pkg/build/controller/build/build_controller.go]: +1

File: pkg/build/controller/build/build_controller.go (hunk 1 above)
Comment (gabemontero) [reply pkg/build/controller/build/build_controller.go]: +1

File: pkg/build/controller/build/build_controller.go (hunk 3 above)
Comment (csrwng) [reply pkg/build/controller/build/build_controller.go]: > Because you know that you didn't successfully complete the work in the current state, therefore you shouldn't advance to the next.

But the work is to move to the next state :)


> The implementation related to handling and advancing the state should be in one place for legibility and maintainability

This I think I agree with. It does allow simpler code.

Comment (gabemontero) [reply pkg/build/controller/build/build_controller.go]: trying to remember the rules for the `workqueue.RateLimitingInterface` ... can we assume we will not get duplicate events?  I *thought* the answer was we could *not* assume that.  If so, we would want this stanza to handle it.

Even if we can assume no duplicate events, this still seems like safe code / protection for something unexpected, including a potential bug in the lister/queue/cache stuff that lead to duplicate events.

File: pkg/build/controller/strategy/util.go (hunk 4 above)
Comment (bparees) [reply pkg/build/controller/strategy/util.go]: it's not so much a common owner ref.  it's one owning the other.  I think the name is fine but if we want something else, "Owns()" would be the name i would suggest.  Or "BuildOwnsPod" is perhaps clearest about the relationship being tested.

**Comment by jim-minter** [comment]
Hmm, I'm not sure if I'm feeling courageous enough to remove the "question 3" stanza.  I fear too much happens under handleNewBuild - e.g. side-effects in different run policies; non-local state in run policies, image and secret resolution.  Maybe that's why this hack exists.

The risk is that some time after a pod is kicked off but the build update fails, we get back into handleNewBuild and make different decisions to those made previously.  The pod is already running but we update the build, or other objects, in a divergent way.

I think we're one state short, and/or we have got other architectural shortcomings.  I'm not convinced it's worth opening this all up.

I think the maximum this PR should do is change the timestamp logic, fix questions 1 & 2, and add documentation around question 3 (assuming others are in agreement with my viewpoint).

Thoughts?

**Comment by gabemontero** [comment]
+1 @jim-minter with some sort of //TODO for your number 3 ... I'm glad I don't have to come up with the wording for that one

**Comment by csrwng** [comment]
Sgtm

**Comment by bparees** [comment]
@jim-minter be aware that we didn't use to set the ownerrefs.  Have you thought through how this impacts someone doing an upgrade who has a build/build-pod that have no owner relationship?

It seems unlikely such a cluster would have a build in the "new" state, but it's not impossible.  You might consider making this check more thorough to detect/handle "older" build pods that have a build annotation, but no ownerref.

**Comment by jim-minter** [comment]
FWIW, I think we've set OwnerRefs since https://github.com/openshift/origin/commit/3091d6aed3bf2b1a4b6c7059de7e6ddf9dfae745, which was in 3.6.

**Comment by bparees** [comment]
/lgtm
/hold
(remove hold when extended builds passes)

**Comment by jim-minter** [comment]
repushed, updated tests only

**Comment by bparees** [comment]
the extended build failure was
```
2018-02-26T19:50:48.083795313Z pulling image error : failed to register layer: devmapper: Error activating devmapper device for 'c4aca1f2288eb8a556d03823c3d8b9d99855e9eec7a9b9448623efadcae03de3': devicemapper: Error running deviceCreate (ActivateDevice) dm_task_run failed
2018-02-26T19:50:48.139010328Z error: build error: unable to get openshift/nodejs-010-centos7@sha256:bd971b467b08b8dbbbfee26bad80dcaa0110b184e0a8dd6c1b0460a6d6f5d332
```

/hold cancel

**Comment by bparees** [comment]
3.9 has been branched
/hold cancel

**Comment by bparees** [comment]
/hold
master is locked
